curl/curl
# TLSv1.3 earlydata support for curl

Based on #14135, implement TLSv1.3 earlydata support for the curl command line, libcurl and its implementation in GnuTLS.

If a known TLS session announces early data support, and the feature is enabled *and* it is not a "connect-only" transfer, delay the TLS handshake until the first request is being sent.

- Add `--tls-earlydata` as new boolean command line option for curl.
- Add `CURLSSLOPT_EARLYDATA` to libcurl to enable use of the feature.
- Add `CURLINFO_EARLYDATA_SENT_T` to libcurl, reporting the amount of bytes sent and accepted/rejected by the server.

Implementation details:
- store the ALPN protocol selected at the SSL session.
- When reusing the session and enabling earlydata, use exactly that ALPN protocol for negotiation with the server. When the sessions ALPN does not match the connections ALPN, earlydata will not be enabled.
- Check that the server selected the correct ALPN protocol for an earlydata connect. If the server does not confirm or reports something different, the connect fails.
- HTTP/2: delay sending the initial SETTINGS frames during connect, if not connect-only.

Verification:
- add test_02_32 to verify earlydata GET with nghttpx.
- add test_07_70 to verify earlydata PUT with nghttpx.
- add support in 'hx-download', 'hx-upload' clients for the feature

/cc @ad-chaos
